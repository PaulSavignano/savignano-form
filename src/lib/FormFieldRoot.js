import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

export class FormFieldRootComponent extends PureComponent {

  componentDidMount() {
    const { fieldProps, onRegisterField } = this.props
    onRegisterField(fieldProps)
  }

  componentWillUnmount() {
    const { isPersistOnUnmount, name, onUnregisterField } = this.props
    if (!isPersistOnUnmount) {
      onUnregisterField({ name })
    }
  }

  render() {
    const {
      component: Comp,
      fieldProps,
      isPersistOnUnmount,
      onFormat,
      onParse,
      onRegisterField,
      onUnregisterField,
      onValidate,
      value,
      ...rest
    } = this.props
    const props = {
      ...rest,
      value: onFormat ? onFormat(value) : value
    }
    return <Comp {...props} />
  }
}

FormFieldRootComponent.defaultProps = {
  id: undefined,
  isPersistOnUnmount: false,
  onFormat: undefined,
  onParse: undefined,
  onValidate: undefined,
  type: 'text',
  value: undefined
}

FormFieldRootComponent.propTypes = {
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.node, PropTypes.string])
    .isRequired,
  fieldProps: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFormat: PropTypes.func,
    onParse: PropTypes.func,
    onValidate: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
    type: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string])
  }).isRequired,
  id: PropTypes.string,
  isPersistOnUnmount: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onFormat: PropTypes.func,
  onParse: PropTypes.func,
  onRegisterField: PropTypes.func.isRequired,
  onUnregisterField: PropTypes.func.isRequired,
  onValidate: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
  type: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string])
}

const Memoized = React.memo(props => <FormFieldRootComponent {...props} />)

export default Memoized
