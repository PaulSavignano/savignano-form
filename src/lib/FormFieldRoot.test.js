import React from 'react'
import { shallow } from 'enzyme'

import { FormFieldRootComponent } from './FormFieldRoot'

const testProps = {
  component: props => <div {...props} />,
  isPersistOnUnmount: false,
  name: 'email',
  onFormat: undefined,
  onParse: undefined,
  onRegisterField: jest.fn(),
  onUnregisterField: jest.fn(),
  onValidate: jest.fn(),
  value: 'testing@test.com'
}

describe('FormFieldRoot', () => {

  it('should match snapshot', () => {
    const wrapper = shallow(<FormFieldRootComponent {...testProps} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should register field on mount', () => {
    const spy = jest.fn()
    const wrapper = shallow(<FormFieldRootComponent {...testProps} onRegisterField={spy} />)
    wrapper.dive()
    expect(spy).toHaveBeenCalled()
  })
  it('should unregister field on unmount if isPersistOnUnmount if false', () => {
    const spy = jest.fn()
    const wrapper = shallow(<FormFieldRootComponent {...testProps} onRegisterField={spy} />)
    wrapper.dive()
    wrapper.instance().componentWillUnmount()
    expect(spy).toHaveBeenCalled()
  })
  it('should not unregister field on unmount if isPersistOnUnmount if true', () => {
    const spy = jest.fn()
    const wrapper = shallow(
      <FormFieldRootComponent
        {...testProps}
        isPersistOnUnmount
        value="testing2@test.com"
        onUnregisterField={spy}
      />
    )
    wrapper.instance().componentWillUnmount()
    expect(spy).not.toHaveBeenCalled()
  })
  it('should call onFormat with value', () => {
    const spy = jest.fn()
    shallow(
      <FormFieldRootComponent
        {...testProps}
        isPersistOnUnmount
        onFormat={spy}
      />
    ).dive()
    expect(spy).toHaveBeenCalledWith(testProps.value)
  })
})
