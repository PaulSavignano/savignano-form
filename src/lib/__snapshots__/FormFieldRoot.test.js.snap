// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldRoot should match snapshot 1`] = `ShallowWrapper {}`;
